### PR TITLE
.NET Glossary

### LINKED ISSUE #1767: Documentation Duplication Between .NET Projects
Porting https://github.com/dotnet/corefx/issues/5899

- Move dotnet/corefx/Documentation/project-docs/glossary.md to dotnet/core-docs
- Follow @cartermp's advice and add new TOC entry "Glossary" under "Understanding the .NET Platform"

@cartermp

### LINKED ISSUE #2574: .NET Glossary
Create a glossary of terms commonly used in .NET docs.  The goal is not to provide a comprehensive list, but to clarify meanings of words that have been misunderstood or cause confusion. A secondary purpose is to provide word use guidance for contributors to follow when editing existing docs or writing new ones.  

See LiveFyre comments on [the Frameworks doc](https://docs.microsoft.com/en-us/dotnet/standard/frameworks) and [the Packages doc](https://docs.microsoft.com/en-us/dotnet/standard/packages)

> I'd really like to see a glossary of the related terms in one spot, including: Framework, Platform, Library, Runtime, etc. The terminology related in relationship .net core is slippery at best and mind-numbing at worst. ...

> As I continue to read on this topic, I'm finding that even within MS, authors are using varying terminology. Can you all get into one room together and decide what constitutes a platform, a package, a runtime, a framework, a library, a metapackage, and so on? A diagram would be with 10,000 words I think.  Somewhere on that diagram I'd like to see some bolded boxes that say "actual physical library code here

Diagrams will be included in the glossary but may not be in its first edition.

There is also a [Reddit thread](https://news.ycombinator.com/item?id=14509339):

> There are terms like .NET Core, .NET Platform, .NET Framework, .NET SDK all mixed up on nearly every page, ...

Proposed location for **.NET Glossary** in the ToC is after **Frameworks and Targets** (which will become **Target Frameworks**):
image

### PR DESCRIPTION
**Pull request by tdykstra**
Fixes #2574 
Fixes #1767 , brings selected entries from the product repo's glossary into this doc

[Review URL](https://review.docs.microsoft.com/en-us/dotnet/standard/glossary?branch=pr-en-us-2676)

### CODE REVIEW

**Comment by tdykstra** [comment]
Review requests were premature -- I have to validate that staging looks OK, especially to see if this large number of h2s renders well.  I'll post another comment here when ready for review.

**Comment by GuardRex** [comment]
Could **WIP** it ... we ignore for review/merge with WIP on it. I've also favored "[WIP]" in the title when setting the label.

**Comment by tdykstra** [comment]
I'll do both, thanks!

**Comment by GuardRex** [comment]
@tdykstra If u want, ping me just before u ping the others for the full review.

I see some low-hanging fruit 🍎 minor/nit changes you'll prob want to make.

**Comment by GuardRex** [comment]
@tdykstra Just a quick nits pass, and I'll read carefully this afternoon.

* Links with *https://docs.microsoft.com/en-us* that should be relative and without `en-us`.
* We use dashes (`-`) over asterisks (`*`) at the first level to create bullet lists in this repo. Then, we use asterisks for the 2nd level. (*Generally speaking!* 😆 lol ... I'm sure there are a bunch of wild formats, but this is the general approach.)
* There are a handful of double-spaces between sentences in the topic; ~10 of them.
* @mairaw: This reminds me to ask u: We're kind'a all over the place (little bit my fault for not asking earlier, too) on referring to files and file extensions. There was already some variability on the repo when I arrived, and I've just been trying to standardize slowly, which won't matter soon. :wink:
  * I've felt it looks better to not include the period when talking about a specific file but leave the extension in italics (e.g., ... the *csproj* project file ...); just because I think it reads better to not have the dot on there (less noisy).
  * When explicitly referring to an extension, I've been using dot notation, which is very common (e.g., ... the *.zip* extension is on the file ...).
  * (At least in a few cases recently) when referring in passing to the entire project system, I haven't been applying style (e.g., When you move your project to the csproj system, you ...).
* In several spots, you place a link after a return (e.g., `See [Target Frameworks](frameworks.md).`), but the preceding line doesn't have two spaces at the end; therefore, there will be no line break when its built. Either use two returns in those spots, or add two spaces, but I think the general way to go when it's not a list is to use two returns.
* *See also* lists: We don't bullet them over here. We do leave them on one line each tho as you have it with no extra lines in between them. In order to make them line break, add two spaces to the end of each line in the list except for the last one.
* Also not 100% sure, but I've received some hate messages on diffs (or someplace) for not putting a CR on the last line. It might be a GH thing ... or maybe it's VSC that bugs me about it ... but most/all of our topics end on a final zero-length line.

I'll take a closer look at content this afternoon, but these were just some quick things that I noticed when scanning the file.

**Comment by tdykstra** [comment]
@GuardRex Let me know when you've had a chance to do your closer look; I'll wait for that before taking down the WIP indicators.

**Comment by GuardRex** [comment]
@tdykstra I'll do it right now ..... I've been mess'in around with the Target Frameworks topic. 😅

File: docs/standard/glossary.md
Comment (GuardRex) [inline docs/standard/glossary.md]: "that is" :point_right: "that's"
"as well as" :point_right: "and"

UWP doesn't actually build headless devices ...

> An implementation of .NET that's used for building modern, touch-enabled Windows applications and software for headless devices used in Internet of Things (IoT) systems.

Comma after VB.NET in ...

> When using C# and VB.NET, the .NET APIs are provided by .NET Core.

File: docs/standard/glossary.md
Comment (GuardRex) [inline docs/standard/glossary.md]: I'd suggest one more CR at the end of this line. It might not matter, but I get :skull: *death threats* :skull: from GH Desktop when I leave a file ending on a full line.

**Review by guardrex — CHANGES_REQUESTED** [review]
Very nice. I like this topic a lot!

File: docs/standard/glossary.md
Comment (GuardRex) [inline docs/standard/glossary.md]: Use relative link

File: docs/standard/glossary.md
Comment (GuardRex) [inline docs/standard/glossary.md]: Use relative link

File: docs/standard/glossary.md
Comment (GuardRex) [inline docs/standard/glossary.md]: Use relative links for this one and the next one.

**Review by guardrex — CHANGES_REQUESTED** [review]
Forgot to mark the links that go relative

File: docs/standard/glossary.md
Comment (tdykstra) [inline docs/standard/glossary.md]: > Why not fix the inconsistent usage? Don't you simply mean that sometimes the context determines what a word/phrase/acronym means?

There are too many articles with too many inconsistent usages to fix them all on anything like a reasonable schedule.  The doc set is too big compared to the writer resources to execute a task like that.

> imply that MS inadvertently uses inconsistent terminology

That's precisely the case

That said, I think we can drop the second sentence and just call out problematic usage in the individual entries.

File: docs/standard/glossary.md
Comment (tdykstra) [inline docs/standard/glossary.md]: In starting with a noun that describes the entry I'm following the lead of the [Azure Glossary](https://docs.microsoft.com/en-us/azure/azure-glossary-cloud-terminology).  I'll leave it now but have asked Monica for the rationale behind the Azure Glossary style.

Putting def on its own line is worth a try.

I don't think we've adopted as a style standard the practice of reducing every occurrence of "application" to "app" but I'll look for additional places where the shortened form might fit better.

Comment (GuardRex) [inline docs/standard/glossary.md]: If there's a strong case for it, it'll prob be that with localization, they'll say that if it isn't a complete sentence that the auto-translation will flake out on a bunch and the translations won't be so hot. Just guessing tho from other comments I've seen regarding how they localize content.

The application to app thing was from the chats I've had mostly with [@}RickAnderson. I try to follow his advice as best I can.

File: docs/standard/glossary.md
Comment (GuardRex) [inline docs/standard/glossary.md]: I'd escape the asterisk on System.* just to be safe. It doesn't seem to confuse the markdown parser tho, so maybe these were ok as they were. Also, consider that normally we inline code these API references, so I wonder if they should be \`System.\*\` for `System.*` (no escape, of course) and \`Microsoft.\*\` for `Microsoft.*`.

File: docs/standard/glossary.md
Comment (GuardRex) [inline docs/standard/glossary.md]: Since there's only one item, do u still want the bullet? ... No hardcore worries from the style manual, of course, as they only dissuade from having this, but it is a "list" of one item. I have to agree with Mrs. Calvary, my 7th grade English teacher, on this one! lol Also because there's only one item, I think the example sentence could be just placed right on the end of the prior sentence in the paragraph. Just a suggestion.

File: docs/standard/glossary.md
Comment (GuardRex) [inline docs/standard/glossary.md]: Dashes here and below for the bullets

**Review by guardrex — APPROVED** [review]
LGTM ... just the one list that needs dashes for bullets. Great work :+1: ... I hope this gets linked into prominent topics. I noticed you don't do that on this PR, but I don't see why u couldn't. It could link into some of the basic, first-read topics that devs access.

**Comment by tdykstra** [comment]
Removing WIP markers, this is ready for review.  Review URL is in the PR description box.

**Comment by tdykstra** [comment]
@mairaw @richlander @BillWagner @rpetrusha @terrajobst 
Luke did a thorough review but I'd like to get at least one more review before merging this

**Comment by GuardRex** [comment]
Indeed! I'm a web dev by trade ... only playing around with CoreRT for fun. I can't quite be trusted on reviewing lingo pertaining to compiler technology ... yet anyway!

**Comment by tdykstra** [comment]
I know some of you are doing OneWeek - is anyone available to review this? @rpetrusha or @terrajobst ?

File: docs/standard/glossary.md
Comment (rpetrusha) [inline docs/standard/glossary.md]: I'd delete "original" -- it seems to me to have an implication of datedness.

File: docs/standard/glossary.md
Comment (rpetrusha) [inline docs/standard/glossary.md]: Maybe expand slightly on removing unused code -- it extracts code from libraries and merges them into the executable as well.

File: docs/standard/glossary.md
Comment (rpetrusha) [inline docs/standard/glossary.md]: Following --> The following

File: docs/standard/glossary.md
Comment (rpetrusha) [inline docs/standard/glossary.md]: I'm not sure it's frequently -- I think that we've tried to use runtime to refer to the common language runtime, though I'm sure there are exceptions. I think that I would change frequently to sometimes.

File: docs/standard/glossary.md
Comment (rpetrusha) [inline docs/standard/glossary.md]: target framework also includes a version. So a target framework is a versioned .NET implementation.

**Review by rpetrusha — CHANGES_REQUESTED** [review]
Sorry this has taken so long, @tdykstra.

File: docs/standard/glossary.md
Comment (tdykstra) [inline docs/standard/glossary.md]: > This section needs some clarification. A framework is different from an application framework. 

The purpose is to start with a general/common meaning of framework -- it's a highly flexible word so I'm open to suggestions here. Ideally it should work in such a way that "application framework" works as a type of framework.

> And I'm not sure that target framework and TFM are necessary here; I think that they add one more confusing detail.

The issue I'm trying to address is confusion that results from use of "framework" with different modifiers that make it have different meanings.  The fact that it is used so differently is not always immediately apparent to readers. In this case I was thinking it's important to point out that "framework" in *target framework* and TFM has a special meaning.

File: docs/standard/glossary.md
Comment (BillWagner) [inline docs/standard/glossary.md]: Somewhere in this paragraph, we need to mention Linux. Suggestion:

"It's still a cross platform runtime." => "It's still a cross platform runtime, now including support for many Linux distributions."

File: docs/standard/glossary.md
Comment (BillWagner) [inline docs/standard/glossary.md]: This seems to conflate namespaces with libraries. I'd suggest changing
"A set of System.* (and to a limited extent Microsoft.*) libraries." to "A set of libraries that comprise the System.* (and to a limited extent Microsoft.*) namespaces."

File: docs/standard/glossary.md
Comment (BillWagner) [inline docs/standard/glossary.md]: Suggestion: cross-link this term and JIT.

File: docs/standard/glossary.md
Comment (BillWagner) [inline docs/standard/glossary.md]: I would think "an implementation" would include *exactly one* runtime.

**Review by BillWagner — APPROVED** [review]
This is a great addition, @tdykstra.

I made a few suggestions, but I say :shipit: when you are ready.

File: docs/standard/glossary.md
Comment (tdykstra) [inline docs/standard/glossary.md]: @terrajobst suggested "one or more".  Maybe an instance of an implementation would only have one? 😄

File: docs/standard/glossary.md
Comment (tdykstra) [inline docs/standard/glossary.md]: @rpetrusha OK to merge as-is, or do you have a suggestion for how to improve this entry?

**Review by rpetrusha — APPROVED** [review]

**Comment by rpetrusha** [comment]
This looks good, @tdykstra. It's ready to merge when you want.

**Pull request merged**
